File: src/mehra/models/providers/llamacpp_provider.py
from typing import Dict, List, Any, Generator
import time
from .model_provider import ModelProvider

try:
    from llama_cpp import Llama
except ImportError:
    raise ImportError(
        "llama-cpp-python is not installed. "
        "Install it with: pip install llama-cpp-python"
    )


class LatencyMetrics:
    """Container for latency measurements."""
    
    def __init__(self):
        self.setup_latency: float = 0.0  # Time to prepare inference (format, params) in ms
        self.first_token_latency: float = 0.0  # Time to first token from model (ms)
        self.time_per_token: float = 0.0  # Average ms per token
        self.total_latency: float = 0.0  # Total inference time (ms)
        self.tokens_generated: int = 0
        self.tokens_per_second: float = 0.0
    
    def __repr__(self) -> str:
        return (
            f"LatencyMetrics("
            f"Setup={self.setup_latency:.2f}ms, "
            f"TTFT={self.first_token_latency:.2f}ms, "
            f"TPS={self.tokens_per_second:.2f} tokens/s, "
            f"Tokens={self.tokens_generated}, "
            f"Total={self.total_latency:.2f}ms)"
        )


class LlamaCppProvider(ModelProvider):
    """Llama.cpp model provider for running GGUF format models locally."""
    
    def __init__(
        self,
        model_path: str,
        n_ctx: int = 2048,
        n_threads: int = 8,
        n_gpu_layers: int = 0,
        verbose: bool = False,
        **kwargs
    ):
        """Initialize the Llama.cpp provider.
        
        Args:
            model_path: Path to the GGUF model file
            n_ctx: Context window size (default: 2048)
            n_threads: Number of threads for inference (default: 8)
            n_gpu_layers: Number of layers to offload to GPU (0 = CPU only, default: 0)
            verbose: Enable verbose output (default: False)
            **kwargs: Additional parameters to pass to Llama
        """
        self.model_path = model_path
        self.n_ctx = n_ctx
        self.n_threads = n_threads
        self.n_gpu_layers = n_gpu_layers
        self.last_metrics: LatencyMetrics = LatencyMetrics()
        
        try:
            self.model = Llama(
                model_path=model_path,
                n_ctx=n_ctx,
                n_threads=n_threads,
                n_gpu_layers=n_gpu_layers,
                verbose=verbose,
                **kwargs
            )
        except Exception as e:
            raise RuntimeError(f"Failed to load model from {model_path}: {str(e)}")
    
    def generate_response(self, messages: List[Dict[str, str]], **kwargs) -> str:
        """Generate a complete response using Llama.cpp.
        
        Args:
            messages: List of message dictionaries with 'role' and 'content'
            **kwargs: Additional parameters
                temperature: Controls randomness (0.0 to 1.0, default: 0.7)
                top_p: Controls diversity (0.0 to 1.0, default: 0.95)
                max_tokens: Maximum number of tokens to generate (default: 512)
        
        Returns:
            Complete text response
        """
        # Collect all chunks from streaming and return as single string
        full_response = ""
        for chunk in self.generate_response_stream(messages, **kwargs):
            full_response += chunk
        return full_response
    
    def generate_response_stream(
        self,
        messages: List[Dict[str, str]],
        **kwargs
    ) -> Generator[str, None, None]:
        """Generate a streaming response using Llama.cpp.
        
        Args:
            messages: List of message dictionaries with 'role' and 'content'
            **kwargs: Additional parameters
                temperature: Controls randomness (0.0 to 1.0, default: 0.7)
                top_p: Controls diversity (0.0 to 1.0, default: 0.95)
                max_tokens: Maximum number of tokens to generate (default: 512)
        
        Yields:
            Text chunks as they are generated
        """
        # Record overall start time (including setup)
        overall_start_time = time.time()
        
        # Extract generation parameters
        temperature = kwargs.get("temperature", 0.7)
        top_p = kwargs.get("top_p", 0.95)
        max_tokens = kwargs.get("max_tokens", 512)
        
        # Format messages into a prompt (setup phase)
        setup_start = time.time()
        prompt = self._format_messages_to_prompt(messages)
        setup_end = time.time()
        
        # Initialize latency tracking
        metrics = LatencyMetrics()
        metrics.setup_latency = (setup_end - setup_start) * 1000
        
        first_token_time = None
        token_count = 0
        model_start_time = setup_end  # When model inference starts

        try:
            # Use llama_cpp's streaming capability
            response = self.model(
                prompt=prompt,
                max_tokens=max_tokens,
                temperature=temperature,
                top_p=top_p,
                stream=True,
            )
            
            # Yield text chunks from the stream
            for chunk in response:
                if "choices" in chunk and len(chunk["choices"]) > 0:
                    delta = chunk["choices"][0].get("text", "")
                    if delta:
                        token_count += 1
                        
                        # Record first token latency (from model start)
                        if first_token_time is None:
                            first_token_time = time.time()
                            metrics.first_token_latency = (first_token_time - model_start_time) * 1000
                        
                        yield delta
                        
            # Calculate final metrics
            end_time = time.time()
            metrics.total_latency = (end_time - overall_start_time) * 1000
            metrics.tokens_generated = token_count
            
            if token_count > 0:
                metrics.time_per_token = (metrics.total_latency - metrics.setup_latency) / token_count
                metrics.tokens_per_second = token_count / ((end_time - model_start_time) or 1)
            
            self.last_metrics = metrics
                        
        except Exception as e:
            yield f"Error: Failed to generate response. {str(e)}"
    
    def get_latency_metrics(self) -> LatencyMetrics:
        """Get the latency metrics from the last inference.
        
        Returns:
            LatencyMetrics object with measurement data
        """
        return self.last_metrics
    
    def _format_messages_to_prompt(self, messages: List[Dict[str, str]]) -> str:
        """Convert messages to a prompt string.
        
        Args:
            messages: List of message dictionaries with 'role' and 'content'
        
        Returns:
            Formatted prompt string
        """
        prompt = ""
        for message in messages:
            role = message.get("role", "user").lower()
            content = message.get("content", "")
            
            if role == "system":
                prompt += f"{content}\n\n"
            elif role == "user":
                prompt += f"User: {content}\n"
            elif role == "assistant":
                prompt += f"Assistant: {content}\n"
        
        # Add prompt for assistant response
        prompt += "Assistant:"
        return prompt
    
    def get_model_info(self) -> Dict[str, Any]:
        """Get information about the loaded model.
        
        Returns:
            Dictionary containing model information
        """
        try:
            return {
                "model_path": self.model_path,
                "context_size": self.n_ctx,
                "threads": self.n_threads,
                "gpu_layers": self.n_gpu_layers,
                "model_loaded": self.model is not None,
            }
        except Exception as e:
            return {"error": str(e)}
    
    def __del__(self):
        """Clean up model resources."""
        if hasattr(self, 'model') and self.model is not None:
            try:
                self.model = None
            except Exception:
                pass

File: src_cs/MeHRa.Models/Providers/LlamaCppProvider.cs
using LLama;
using LLama.Common;
using LLama.Native;

namespace MeHRa.Models.Providers;

/// <summary>
/// Represents latency metrics for inference operations.
/// </summary>
public class LatencyMetrics
{
    /// <summary>
    /// Time to prepare inference (format prompt, setup params) in milliseconds.
    /// </summary>
    public double SetupLatency { get; set; }

    /// <summary>
    /// Time to first token from model in milliseconds.
    /// </summary>
    public double FirstTokenLatency { get; set; }

    /// <summary>
    /// Average time per token in milliseconds.
    /// </summary>
    public double TimePerToken { get; set; }

    /// <summary>
    /// Total inference time in milliseconds (including setup).
    /// </summary>
    public double TotalLatency { get; set; }

    /// <summary>
    /// Number of tokens generated.
    /// </summary>
    public int TokensGenerated { get; set; }

    /// <summary>
    /// Tokens generated per second.
    /// </summary>
    public double TokensPerSecond { get; set; }

    public override string ToString()
    {
        return $"LatencyMetrics(" +
               $"Setup={SetupLatency:F2}ms, " +
               $"TTFT={FirstTokenLatency:F2}ms, " +
               $"TPS={TokensPerSecond:F2} tokens/s, " +
               $"Tokens={TokensGenerated}, " +
               $"Total={TotalLatency:F2}ms)";
    }
}

/// <summary>
/// Llama.cpp model provider for running GGUF format models locally using LLamaSharp.
/// </summary>
public class LlamaCppProvider : ModelProvider, IAsyncDisposable
{
    private readonly string _modelPath;
    private readonly int _contextSize;
    private readonly int _threads;
    private readonly int _gpuLayers;
    private LLamaWeights? _model;
    private LLamaContext? _context;
    private LatencyMetrics _lastMetrics = new();

    /// <summary>
    /// Initialize the Llama.cpp provider.
    /// </summary>
    /// <param name="modelPath">Path to the GGUF model file</param>
    /// <param name="contextSize">Context window size (default: 2048)</param>
    /// <param name="threads">Number of threads for inference (default: 8)</param>
    /// <param name="gpuLayers">Number of layers to offload to GPU (0 = CPU only, default: 0)</param>
    /// <exception cref="FileNotFoundException">Thrown if model file is not found</exception>
    /// <exception cref="InvalidOperationException">Thrown if model loading fails</exception>
    public LlamaCppProvider(
        string modelPath,
        int contextSize = 2048,
        int threads = 8,
        int gpuLayers = 0)
    {
        if (!File.Exists(modelPath))
        {
            throw new FileNotFoundException($"Model file not found: {modelPath}");
        }

        _modelPath = modelPath;
        _contextSize = contextSize;
        _threads = threads;
        _gpuLayers = gpuLayers;

        LoadModel();
    }

    /// <summary>
    /// Load the model from the specified path.
    /// </summary>
    /// <exception cref="InvalidOperationException">Thrown if model loading fails</exception>
    private void LoadModel()
    {
        try
        {
            var parameters = new ModelParams(_modelPath)
            {
                ContextSize = (uint)_contextSize,
                ThreadCount = _threads,
                GpuLayerCount = _gpuLayers,
                Verbose = false,
            };

            _model = LLamaWeights.LoadFromFile(parameters);
            _context = _model.CreateContext(parameters);
        }
        catch (Exception ex)
        {
            throw new InvalidOperationException($"Failed to load model from {_modelPath}: {ex.Message}", ex);
        }
    }

    /// <summary>
    /// Generate a complete response using Llama.cpp.
    /// </summary>
    /// <param name="messages">List of message objects with role and content</param>
    /// <param name="cancellationToken">Cancellation token for async operations</param>
    /// <returns>Complete text response</returns>
    public override async Task<string> GenerateResponseAsync(
        List<Message> messages,
        CancellationToken cancellationToken = default)
    {
        var response = "";
        await foreach (var chunk in GenerateResponseStreamAsync(messages, cancellationToken))
        {
            response += chunk;
        }
        return response;
    }

    /// <summary>
    /// Generate a streaming response using Llama.cpp.
    /// </summary>
    /// <param name="messages">List of message objects with role and content</param>
    /// <param name="cancellationToken">Cancellation token for async operations</param>
    /// <returns>Async enumerable of text chunks</returns>
    public override async IAsyncEnumerable<string> GenerateResponseStreamAsync(
        List<Message> messages,
        CancellationToken cancellationToken = default)
    {
        if (_model == null || _context == null)
        {
            throw new InvalidOperationException("Model not loaded. Call LoadModel first.");
        }

        // Record overall start time (including setup)
        var overallStart = DateTime.UtcNow;

        // Setup phase: format messages to prompt
        var setupStart = DateTime.UtcNow;
        var prompt = FormatMessagesToPrompt(messages);
        var setupEnd = DateTime.UtcNow;

        // Initialize latency tracking
        var metrics = new LatencyMetrics();
        metrics.SetupLatency = (setupEnd - setupStart).TotalMilliseconds;

        var executor = new StatelessExecutor(_model, _context.Params);
        var inferenceParams = new InferenceParams
        {
            Temperature = 0.7f,
            TopP = 0.95f,
            AntiPrompts = new List<string> { "User:", "Assistant:" },
        };

        DateTime? firstTokenTime = null;
        var tokenCount = 0;
        var modelStartTime = setupEnd;

        await foreach (var token in executor.InferAsync(prompt, inferenceParams, cancellationToken))
        {
            tokenCount++;

            // Record first token latency (from model start)
            if (firstTokenTime == null)
            {
                firstTokenTime = DateTime.UtcNow;
                metrics.FirstTokenLatency = (firstTokenTime.Value - modelStartTime).TotalMilliseconds;
            }

            yield return token;
        }

        // Calculate final metrics
        var endTime = DateTime.UtcNow;
        metrics.TotalLatency = (endTime - overallStart).TotalMilliseconds;
        metrics.TokensGenerated = tokenCount;

        if (tokenCount > 0)
        {
            var modelDuration = (endTime - modelStartTime).TotalSeconds;
            metrics.TimePerToken = (metrics.TotalLatency - metrics.SetupLatency) / tokenCount;
            metrics.TokensPerSecond = modelDuration > 0 ? tokenCount / modelDuration : 0;
        }

        _lastMetrics = metrics;
    }

    /// <summary>
    /// Convert messages to a prompt string.
    /// </summary>
    /// <param name="messages">List of message objects with role and content</param>
    /// <returns>Formatted prompt string</returns>
    private string FormatMessagesToPrompt(List<Message> messages)
    {
        var prompt = "";

        foreach (var message in messages)
        {
            var role = message.Role.ToLower();

            prompt += role switch
            {
                "system" => $"{message.Content}\n\n",
                "user" => $"User: {message.Content}\n",
                "assistant" => $"Assistant: {message.Content}\n",
                _ => $"{role}: {message.Content}\n"
            };
        }

        prompt += "Assistant:";
        return prompt;
    }

    /// <summary>
    /// Get information about the loaded model.
    /// </summary>
    /// <returns>Dictionary containing model information</returns>
    public Dictionary<string, object> GetModelInfo()
    {
        return new Dictionary<string, object>
        {
            { "ModelPath", _modelPath },
            { "ContextSize", _contextSize },
            { "Threads", _threads },
            { "GpuLayers", _gpuLayers },
            { "ModelLoaded", _model != null },
        };
    }

    /// <summary>
    /// Get the latency metrics from the last inference.
    /// </summary>
    /// <returns>LatencyMetrics object with measurement data</returns>
    public LatencyMetrics GetLatencyMetrics()
    {
        return _lastMetrics;
    }

    /// <summary>
    /// Dispose of model resources.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        _context?.Dispose();
        _model?.Dispose();
        await Task.CompletedTask;
    }
}
